File: JAVA/Clients/clients/src/main/java/com/futureprograms/clients/service/ImageService.java
package com.futureprograms.clients.service;

import com.futureprograms.clients.enums.Gender;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;
import org.springframework.web.multipart.MultipartFile;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.io.IOException;

@Service
@Slf4j
public class ImageService {

    @Value("${upload.dir:./uploads}")
    private String uploadDir;

    public String getUploadDir() {
        // The uploadDir property from application.properties
        return uploadDir;
    }

    /**
     * Resuelve el directorio base real donde se guardan las imágenes.
     * Si la ruta configurada es relativa, se ancla al directorio de ejecución.
     */
    private Path resolveUploadBasePath() {
        Path basePath = Paths.get(uploadDir);
        if (!basePath.isAbsolute()) {
            basePath = Paths.get(System.getProperty("user.dir")).resolve(basePath);
        }
        return basePath.normalize();
    }

    /**
     * Asegura que exista la carpeta base y la carpeta del usuario.
     */
    public Path ensureUserImageDirectory(Long userId) throws IOException {
        Path uploadPath = resolveUploadBasePath();
        Files.createDirectories(uploadPath);

        Path userPath = uploadPath.resolve(String.valueOf(userId));
        Files.createDirectories(userPath);

        return userPath;
    }

    /**
     * Verifica si una imagen es protegida (por defecto según género)
     */
    public boolean isProtectedImage(String imagePath) {
        if (imagePath == null || imagePath.trim().isEmpty()) {
            return false;
        }
        return Gender.isDefaultImagePath(imagePath);
    }

    /**
     * Obtiene la ruta de imagen por defecto para un género
     */
    public String getDefaultImagePath(String genderDisplayName) {
        try {
            Gender gender = Gender.fromDisplayName(genderDisplayName);
            return gender.getDefaultImagePath();
        } catch (IllegalArgumentException e) {
            log.error("Gender no válido: {}", genderDisplayName);
            throw new RuntimeException("Género no válido");
        }
    }

    /**
     * Elimina una imagen (solo si no es protegida)
     */
    public boolean deleteImage(String imagePath) {
        // Validar que no es imagen protegida
        if (isProtectedImage(imagePath)) {
            log.warn("Intento de eliminar imagen protegida: {}", imagePath);
            throw new RuntimeException("No se pueden eliminar imágenes por defecto");
        }

        try {
            Path path = Paths.get(uploadDir, imagePath);
            if (Files.exists(path)) {
                Files.delete(path);
                log.info("Imagen eliminada: {}", imagePath);
                return true;
            }
            log.warn("Imagen no encontrada: {}", imagePath);
            return false;
        } catch (Exception e) {
            log.error("Error al eliminar imagen: {}", imagePath, e);
            throw new RuntimeException("Error al eliminar imagen: " + e.getMessage());
        }
    }

    /**
     * Verifica si un archivo de imagen existe
     */
    public boolean imageExists(String imagePath) {
        try {
            Path path = Paths.get(uploadDir, imagePath);
            return Files.exists(path);
        } catch (Exception e) {
            log.error("Error al verificar imagen: {}", imagePath, e);
            return false;
        }
    }

    /**
     * Valida si una ruta es válida (no contiene caracteres peligrosos)
     */
    public boolean isValidImagePath(String imagePath) {
        if (imagePath == null || imagePath.trim().isEmpty()) {
            return false;
        }
        // Evitar path traversal
        return !imagePath.contains("..") && !imagePath.contains("//") && 
               imagePath.matches("^[a-zA-Z0-9/_.-]+$");
    }

    /**
     * Obtiene el nombre de archivo de una ruta
     */
    public String getImageFileName(String imagePath) {
        if (imagePath == null) {
            return null;
        }
        return imagePath.substring(imagePath.lastIndexOf("/") + 1);
    }

    /**
     * Genera una ruta segura para una imagen
     */
    public String generateSecureImagePath(String fileName, Long userId) {
        // Validar nombre de archivo
        if (!fileName.matches("^[a-zA-Z0-9._-]+$")) {
            throw new RuntimeException("Nombre de archivo inválido");
        }

        // Crear carpeta por usuario
        String userImageDir = String.format("users/%d/", userId);
        return userImageDir + System.currentTimeMillis() + "_" + fileName;
    }

    /**
     * Guarda una imagen de perfil de usuario con nombre fijo "profile"
     */
    public String saveProfileImage(MultipartFile file, Long userId) throws Exception {
        if (file == null || file.isEmpty()) {
            throw new RuntimeException("El archivo de imagen está vacío");
        }

        // Validar tipo de archivo
        String contentType = file.getContentType();
        if (!isValidImageType(contentType)) {
            throw new RuntimeException("Tipo de archivo no permitido. Solo se permiten imágenes (jpg, png, gif, webp)");
        }

        // Obtener extensión original
        String originalFilename = file.getOriginalFilename();
        String extension = "jpg"; // default
        if (originalFilename != null && originalFilename.contains(".")) {
            extension = originalFilename.substring(originalFilename.lastIndexOf(".") + 1).toLowerCase();
        }

        try {
            // Verificar y crear el directorio padre y la carpeta del usuario si no existen
            Path userPath = ensureUserImageDirectory(userId);
            log.info("📁 Carpeta lista para usuario ID {}: {}", userId, userPath.toAbsolutePath());

            // El nombre siempre será "profile" con su extensión original
            String fileName = "profile." + extension;
            Path filePath = userPath.resolve(fileName);

            // Guardar archivo (sobrescribirá si ya existe)
            log.info("📝 Guardando archivo en: {}", filePath.toAbsolutePath());
            Files.write(filePath, file.getBytes());
            log.info("✅ Archivo guardado correctamente. Tamaño: {} bytes", file.getSize());
            
            // La ruta que guardamos en BD será relativa a uploadDir: "ID/profile.ext"
            String dbPath = userId + "/" + fileName;
            log.info("📸 Imagen de perfil guardada para usuario {}: {} (Ruta absoluta: {})", userId, dbPath, filePath.toAbsolutePath());
            
            return dbPath;
        } catch (Exception e) {
            log.error("❌ Error al guardar imagen para usuario {}: {} | Directorio configurado: {} | Causa: {}", 
                    userId, e.getMessage(), resolveUploadBasePath(), e.getClass().getSimpleName(), e);
            throw new RuntimeException("Error al guardar imagen: " + e.getMessage() + " (Verifica que el directorio " + resolveUploadBasePath() + " sea accesible)");
        }
    }

    /**
     * Guarda una imagen de perfil de usuario (MÉTODO ANTIGUO - DEPRECADO)
     */
    @Deprecated
    public String saveProfileImage(MultipartFile file) throws Exception {
        return saveProfileImage(file, System.currentTimeMillis());
    }

    /**
     * Elimina la imagen de perfil de un usuario específico
     * - Si es imagen protegida (default), no elimina nada
     * - Si es imagen personalizada, elimina todo el directorio del usuario
     */
    public void deleteProfileImage(Long userId, String imagePath) {
        if (imagePath == null || imagePath.isEmpty()) {
            log.warn("Ruta de imagen vacía para usuario: {}", userId);
            return;
        }

        // Si es imagen protegida, no hacer nada
        if (isProtectedImage(imagePath)) {
            log.info("📸 Imagen protegida (por defecto) para usuario {}. No se eliminará.", userId);
            return;
        }

        // Eliminar toda la carpeta del usuario
        deleteUserProfileImages(userId);
    }

    /**
     * Elimina todos los archivos de perfil de un usuario (carpeta completa)
     * Se usa cuando se elimina la cuenta del usuario
     */
    public void deleteUserProfileImages(Long userId) {
        try {
            Path userPath = resolveUploadBasePath().resolve(userId.toString());
            
            if (!Files.exists(userPath)) {
                log.warn("📁 Carpeta de usuario no encontrada: {}", userPath);
                return;
            }

            // Eliminar recursivamente toda la carpeta
            Files.walk(userPath)
                    .sorted((a, b) -> b.compareTo(a)) // Ordenar en reverso para eliminar primero archivos
                    .forEach(path -> {
                        try {
                            Files.delete(path);
                            log.info("✅ Eliminado: {}", path);
                        } catch (Exception e) {
                            log.warn("⚠️ No se pudo eliminar: {} - {}", path, e.getMessage());
                        }
                    });

            log.info("✅ Carpeta de imágenes del usuario {} eliminada completamente", userId);
        } catch (Exception e) {
            log.error("❌ Error al eliminar carpeta de imágenes del usuario {}: {}", userId, e.getMessage());
            // No lanzar excepción, solo logging para no interrumpir la eliminación del usuario
        }
    }

    /**
     * Valida el tipo de contenido de la imagen
     */
    private boolean isValidImageType(String contentType) {
        if (contentType == null) {
            return false;
        }
        return contentType.equals("image/jpeg") ||
               contentType.equals("image/png") ||
               contentType.equals("image/gif") ||
               contentType.equals("image/webp");
    }
}

File: ASP .NET/Clients/Services/ImageService.cs
namespace Clients.Services;

using Clients.Enums;

public interface IImageService
{
    Task<string> SaveProfileImageAsync(IFormFile file, long userId);
    Task DeleteProfileImageAsync(string? fileName);
    string GetUploadDirectory();
    bool IsProtectedImage(string fileName);
    Task DeleteImageAsync(string fileName);
    Task DeleteUserProfileImagesAsync(long userId);
    bool IsValidImagePath(string imagePath);
    bool ImageExists(string imagePath);
    string GetImageFileName(string imagePath);
    string GenerateSecureImagePath(string fileName, long userId);
    bool IsValidImageType(string? contentType);
    Task<string> SaveProfileImageAsync(IFormFile file, long userId, Gender gender);
}

public class ImageService : IImageService
{
    private readonly IConfiguration _configuration;
    private readonly ILogger<ImageService> _logger;

    public ImageService(IConfiguration configuration, ILogger<ImageService> logger)
    {
        _configuration = configuration;
        _logger = logger;
    }

    /// <summary>
    /// Convierte una ruta potencialmente relativa a una ruta absoluta
    /// </summary>
    private string GetAbsolutePath(string? path)
    {
        if (string.IsNullOrEmpty(path))
            path = "uploads";

        // Si ya es una ruta absoluta, devolverla tal cual
        if (Path.IsPathRooted(path))
            return path;

        // Si es relativa, combinarla con el directorio actual
        return Path.Combine(Directory.GetCurrentDirectory(), path);
    }

    /// <summary>
    /// Guarda una imagen de perfil con nombre fijo "profile" para un usuario
    /// </summary>
    public async Task<string> SaveProfileImageAsync(IFormFile file, long userId)
    {
        if (file == null || file.Length == 0)
            throw new ArgumentException("El archivo de imagen está vacío");

        // Validar tipo de archivo
        if (!IsValidImageType(file.ContentType))
            throw new ArgumentException("Tipo de archivo no permitido. Solo se permiten imágenes (jpg, png, gif, webp)");

        var configuredUploadDir = _configuration["FileUpload:UploadDirectory"];
        var uploadDir = GetAbsolutePath(configuredUploadDir);

        try
        {
            if (!Directory.Exists(uploadDir))
            {
                Directory.CreateDirectory(uploadDir);
                _logger.LogInformation($"📁 Carpeta de uploads creada: {uploadDir}");
            }
        }
        catch (Exception ex)
        {
            _logger.LogError($"❌ Error al crear directorio de uploads: {ex.Message}");
            throw new InvalidOperationException($"No se puede crear la carpeta de uploads: {ex.Message}", ex);
        }

        try
        {
            // Obtener extensión original
            string originalFilename = file.FileName;
            string extension = "jpg"; // default
            if (!string.IsNullOrEmpty(originalFilename) && originalFilename.Contains("."))
            {
                extension = originalFilename.Substring(originalFilename.LastIndexOf(".") + 1).ToLower();
            }

            // Crear directorio del usuario si no existe
            string userDir = Path.Combine(uploadDir, userId.ToString());
            if (!Directory.Exists(userDir))
            {
                try
                {
                    Directory.CreateDirectory(userDir);
                    _logger.LogInformation($"📁 Carpeta de usuario creada: {userDir}");
                }
                catch (Exception dirEx)
                {
                    _logger.LogError($"❌ Error al crear carpeta del usuario: {dirEx.Message}");
                    throw new InvalidOperationException($"No se puede crear la carpeta del usuario: {dirEx.Message}", dirEx);
                }
            }

            // El nombre siempre será "profile" con su extensión original
            string fileName = $"profile.{extension}";
            string filePath = Path.Combine(userDir, fileName);

            // Guardar archivo (sobrescribirá si ya existe)
            try
            {
                using (var stream = new FileStream(filePath, FileMode.Create))
                {
                    await file.CopyToAsync(stream);
                }
                _logger.LogInformation($"📸 Imagen guardada exitosamente en: {filePath}");
            }
            catch (Exception fileEx)
            {
                _logger.LogError($"❌ Error al guardar archivo: {fileEx.Message}");
                throw new InvalidOperationException($"No se puede guardar la imagen: {fileEx.Message}", fileEx);
            }

            // La ruta que guardamos en BD será relativa a uploadDir: "ID/profile.ext"
            string dbPath = $"{userId}/{fileName}";
            _logger.LogInformation($"📸 Imagen de perfil guardada para usuario {userId}: {dbPath}");
            return dbPath;
        }
        catch (InvalidOperationException)
        {
            throw;
        }
        catch (Exception ex)
        {
            _logger.LogError($"❌ Error al guardar imagen para usuario {userId}: {ex.Message}");
            throw;
        }
    }

    /// <summary>
    /// Guarda una imagen de perfil asignando automáticamente la imagen por defecto si no se proporciona
    /// </summary>
    public async Task<string> SaveProfileImageAsync(IFormFile file, long userId, Gender gender)
    {
        if (file == null || file.Length == 0)
        {
            // Retornar imagen por defecto
            string defaultImage = gender.GetDefaultImagePath();
            _logger.LogInformation($"📸 Asignando imagen por defecto para usuario {userId}: {defaultImage}");
            return defaultImage;
        }

        return await SaveProfileImageAsync(file, userId);
    }

    public async Task DeleteProfileImageAsync(string? fileName)
    {
        if (string.IsNullOrEmpty(fileName))
            return;

        var configuredUploadDir = _configuration["FileUpload:UploadDirectory"];
        var uploadDir = GetAbsolutePath(configuredUploadDir);
        var filePath = Path.Combine(uploadDir, fileName);

        try
        {
            if (File.Exists(filePath))
            {
                File.Delete(filePath);
                _logger.LogInformation($"✅ Imagen eliminada: {fileName}");
            }
            else
            {
                _logger.LogWarning($"⚠️ Archivo no encontrado para eliminar: {filePath}");
            }
        }
        catch (Exception ex)
        {
            _logger.LogError($"❌ Error al eliminar imagen: {ex.Message}");
        }
    }

    public string GetUploadDirectory()
    {
        var configuredUploadDir = _configuration["FileUpload:UploadDirectory"];
        return GetAbsolutePath(configuredUploadDir);
    }

    public bool IsProtectedImage(string? fileName)
    {
        if (string.IsNullOrEmpty(fileName))
            return false;

        return GenderExtensions.IsDefaultImagePath(fileName);
    }

    public async Task DeleteImageAsync(string fileName)
    {
        // No eliminar imágenes protegidas
        if (IsProtectedImage(fileName))
        {
            _logger.LogWarning($"⚠️ Intento de eliminar imagen protegida: {fileName}");
            throw new InvalidOperationException("No se pueden eliminar imágenes por defecto");
        }

        var configuredUploadDir = _configuration["FileUpload:UploadDirectory"];
        var uploadDir = GetAbsolutePath(configuredUploadDir);
        var filePath = Path.Combine(uploadDir, fileName);

        try
        {
            if (File.Exists(filePath))
            {
                File.Delete(filePath);
                _logger.LogInformation($"✅ Imagen eliminada: {fileName}");
            }
            else
            {
                _logger.LogWarning($"⚠️ Imagen no encontrada: {filePath}");
            }
        }
        catch (Exception ex)
        {
            _logger.LogError($"❌ Error al eliminar imagen: {ex.Message}");
            throw;
        }
    }

    /// <summary>
    /// Elimina todos los archivos de perfil de un usuario (carpeta completa)
    /// Se usa cuando se elimina la cuenta del usuario
    /// </summary>
    public async Task DeleteUserProfileImagesAsync(long userId)
    {
        try
        {
            var configuredUploadDir = _configuration["FileUpload:UploadDirectory"];
            var uploadDir = GetAbsolutePath(configuredUploadDir);
            string userPath = Path.Combine(uploadDir, userId.ToString());

            if (!Directory.Exists(userPath))
            {
                _logger.LogWarning($"📁 Carpeta de usuario no encontrada: {userPath}");
                return;
            }

            // Eliminar recursivamente toda la carpeta
            var files = Directory.GetFiles(userPath);
            foreach (var file in files)
            {
                File.Delete(file);
                _logger.LogInformation($"✅ Archivo eliminado: {file}");
            }

            // Eliminar carpeta
            Directory.Delete(userPath);
            _logger.LogInformation($"✅ Carpeta de imágenes del usuario {userId} eliminada completamente");
        }
        catch (Exception ex)
        {
            _logger.LogError($"❌ Error al eliminar carpeta de imágenes del usuario {userId}: {ex.Message}");
            // No lanzar excepción para no interrumpir la eliminación del usuario
        }
    }

    public bool IsValidImagePath(string? imagePath)
    {
        if (string.IsNullOrEmpty(imagePath))
            return false;

        // Evitar path traversal
        return !imagePath.Contains("..") && !imagePath.Contains("//") &&
               System.Text.RegularExpressions.Regex.IsMatch(imagePath, @"^[a-zA-Z0-9/_.-]+$");
    }

    public bool ImageExists(string? imagePath)
    {
        try
        {
            if (string.IsNullOrEmpty(imagePath))
                return false;

            var configuredUploadDir = _configuration["FileUpload:UploadDirectory"];
            var uploadDir = GetAbsolutePath(configuredUploadDir);
            string filePath = Path.Combine(uploadDir, imagePath);
            return File.Exists(filePath);
        }
        catch (Exception ex)
        {
            _logger.LogError($"❌ Error al verificar imagen: {ex.Message}");
            return false;
        }
    }

    public string GetImageFileName(string? imagePath)
    {
        if (string.IsNullOrEmpty(imagePath))
            return "";

        return imagePath.Substring(imagePath.LastIndexOf("/") + 1);
    }

    public string GenerateSecureImagePath(string fileName, long userId)
    {
        // Validar nombre de archivo
        if (!System.Text.RegularExpressions.Regex.IsMatch(fileName, @"^[a-zA-Z0-9._-]+$"))
            throw new ArgumentException("Nombre de archivo inválido");

        // Crear ruta con timestamp para evitar colisiones
        string userImageDir = $"users/{userId}/";
        return userImageDir + DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() + "_" + fileName;
    }

    public bool IsValidImageType(string? contentType)
    {
        if (string.IsNullOrEmpty(contentType))
            return false;

        return contentType switch
        {
            "image/jpeg" => true,
            "image/png" => true,
            "image/gif" => true,
            "image/webp" => true,
            _ => false
        };
    }
}
